backends/bmv2/simple_switch/midend.cpp

```diff
@@ -125,9 +126,10 @@ SimpleSwitchMidEnd::SimpleSwitchMidEnd(CompilerOptions& options, std::ostream* o
             // control plane API, we remove them as well for P4-14 programs.
             isv1 ? new P4::RemoveUnusedActionParameters(&refMap) : nullptr,
             new P4::TypeChecking(&refMap, &typeMap),
-            new P4::MidEndLast(),
             evaluator,
             [this, evaluator]() { toplevel = evaluator->getToplevelBlock(); },
+            options.loopsUnrolling ? new P4::ParsersUnroll(true, &refMap, &typeMap) : nullptr,
```

Comment:
I am worried about this being after the evaluator. Why can't it be before?

Reply:
I think this is a bug, evalutor can't be used after P4::ParsersUnroll. After P4::ParsersUnroll the evalutor (P4::TypeChecking) generates the following error:
../testdata/p4_16_samples/parser-unroll-test2.p4(74): [--Werror=type-error] error: packet.extract
        packet.extract(hdr.srcRoutes.next);
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ---- Actual error:
  ../testdata/p4_16_samples/parser-unroll-test2.p4(74): : Read-only value used for out/inout parameter 'hdr'
          packet.extract(hdr.srcRoutes.next);
                         ^^^^^^^^^^^^^
  ../p4include/core.p4(38)
      void extract<T>(out T hdr);
                            ^^^
  ---- Originating from:
  ../p4include/core.p4(38): Function type 'extract' does not match invocation type '<Method call>'
      void extract<T>(out T hdr);
           ^^^^^^^
  ../testdata/p4_16_samples/parser-unroll-test2.p4(74)
          packet.extract(hdr.srcRoutes.next);

I think that after rewriting we need to update some data structures, but for now we don't known which data structures should be updated. If you have any suggestions, please, let us known. In any case we will continue our analysis.

Reply:
Can you please rebase again against master (soon to be renamed to main?)
This should resolve the problem with gtest.
We have to figure out what the problem is with the evaluator. So you should really run unrolling before the evaluator, dump the produced P4 program using the --top4 flag and check whether hdr.srcRoutes is read-only in some of the generated code. The error is signaled in the source code, but the internal code maintains references to the original sources, so you will have to figure out which statement attempts to write to a read-only value. You can do that by running under the debugger and printing the statement where the error is signaled. There's a dumpP4 utility function for that purpose, or a function called dump.

Reply:
Finally, we got a solution. The problem was [here](https://github.com/Private-Enterprise-LitSoft/p4c/blob/4afbf650c4021525f4d4545b231ab331dea210c3/midend/parserUnroll.cpp#L164). We have removed "typeMap->setType(newArray, expression->expr->type);" and all tests were passed. We have added the tests for unrolling in a separate folder, because "--loopsUnrol" is optional and all p4 files are added as the test cases from the folder [here](https://github.com/Private-Enterprise-LitSoft/p4c/blob/7f9c856d8f4a7e72d18a096c5f56f97abb510c77/backends/p4test/CMakeLists.txt#L119).